uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php

```diff
@@ -38,118 +39,116 @@ public static function create(&$params) {
-      . "WHERE DATE(ac.expire_date) <= %1 AND ac.clone_date IS NULL AND a.is_deleted = 0 AND a.status_id = 3 AND ov.component_id = c.id AND a.activity_type_id = ov.value";
-    $params = array(
-      1 => array($expireDate->format('Y-m-d'), 'String'),
-    );
-    $result = CRM_Core_DAO::executeQuery($query, $params);
-    while ($result->fetch()) {
-      $document = civicrm_api3('Document', 'getsingle', array(
-        'id' => $result->entity_id,
-        'return' => 'activity_type_id, activity_date_time, details, status_id, is_deleted, expire_date, case_id, source_contact_id, target_contact_id, assignee_contact_id',
-      ));
-      $clone = self::cloneDocument($document, $result->id);
-      if (!empty($clone['id'])) {
-        $count++;
-      }
+  protected static function cloneDocument(array $original) {
+    $clone = $original;
+    $origId = $original['id'];
+
+    $clone['activity_date_time'] = $original['expire_date'];
+    $clone['status_id'] = self::STATUS_AWAITING_UPLOAD;
+
+    $fieldsToUnset = ['id', 'file_count', 'expire_date', 'document_number', 'valid_from'];
+    foreach ($fieldsToUnset as $field) {
+      unset($clone[$field]);
     }
-    return $count;
+
+    civicrm_api3('Document', 'create', $clone);
+
+    // Update original's clone_date' so we know it has been cloned.
+    $params = ['id' => $origId, 'clone_date' => date('Y-m-d')];
+    civicrm_api3('Document', 'create', $params);
   }
 
   /**
-   * Clone particular document but with unsetting some fields according to
-   * PCHR-1365 requirements.
-   * 
-   * @param array $document
-   * @param int $expireId
-   * @return array
+   * Gets the custom field name for Documents
+   *
+   * @param string $fieldName
+   *  The human-readable name for the field, e.g. "expiry_date"
+   *
+   * @return string|null
    */
-  protected static function cloneDocument(array $document, $expireId) {
-    if (empty($document['id'])) {
-      return NULL;
-    }
-    $originalId = $document['id'];
-    $expireDateField = self::getExpireDateCustomFieldName();
-    $document['activity_date_time'] = $document[$expireDateField];
-    $document['status_id'] = self::STATUS_AWAITING_UPLOAD;
-    unset($document[$expireDateField]);
-    unset($document[$expireDateField . '_' . $expireId]);
-    unset($document['expire_date']);
-    unset($document['file_count']);
-    unset($document['id']);
-    $clone = civicrm_api3('Document', 'create', $document);
-    if (!empty($clone['id'])) {
-      // Update original document's 'clone_date' with current date
-      // so so we'll know that the document has been already cloned.
-      civicrm_api3('Document', 'create', array(
-        'id' => $originalId,
-        'clone_date' => (new DateTime())->format('Y-m-d'),
-      ));
+  private static function getCustomFieldName($fieldName) {
+    static $fieldNames;
+
+    if (empty($fieldNames)) {
+      $result = civicrm_api3('Document', 'getcustomfields');
```

Comment:
@davialexandre I also made changes here since your review. I think this is simpler than those fancy array functions